### PR TITLE
Modify PyROS Master Feasibility Problem

### PR DESCRIPTION
**Pull request by shermanjasonaf**
## Summary/Motivation:
With the aim of improving the computational performance and convergence guarantees of PyROS, we change the formulation of the master feasibility problem, used to initialize variables for the original master model, from
image
to
image
in which:
- *x* denotes the model's first stage variables
- *z^k* denotes the second-stage variables corresponding to iteration _k_
- *y^k* denotes the second-stage variables corresponding to iteration _k_
- *q^k* denotes the uncertain parameter realization obtained from iteration _k_ (and _q^0_ is the nominal realization)
- *d_l* denotes the decision rule variables corresponding to the *l*th dimension of the 2nd stage var space
- _s_ denotes slack variables 
- *c* denotes slack variable  scaling coefficients
- _K_ denotes the iteration set
- _Q_ denotes the uncertainty set

In particular, we add slack variables to the master model constraints corresponding to the uncertain parameter realization derived from the previous iteration only (as opposed to all master model constraints).

Model variables are meant to be initialized as follows:
image
in which the alphas are auxiliary values used to initialize the slack variables.
For each constraint, the corresponding slack variables _s^(+/-/obj/ineq)_ are initialized to the value _α^(+/-/obj/ineq)_. We set all coefficients _c_ to 1.

Moreover,
- No master feasibility problem is solved in the first PyROS iteration (iteration 0). Master problem variables are initialized to the user-provided values.
- After the master feasibility problem is solved, the master problem variables are initialized to their master feasibility solution counterparts regardless of the slack objective function value (currently, the initialization is only performed if the slack objective is 0).  

## Changes proposed in this PR:
- Modify PyROS master feasibility problem as described above
- Update version number
- Add changelog file (`pyomo/contrib/pyros/CHANGELOG.txt`).

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by emma58 — COMMENTED** [review]
It looks like there is one place that you need to make sure you collect Vars in a deterministic order. Otherwise this looks good to me!

File: pyomo/contrib/pyros/master_problem_methods.py [hunk 1]
```diff
@@ -37,35 +41,186 @@ def initial_construct_master(model_data):
...
+    for itn in iterations:
+        fsv_set = ComponentSet(
+                model.scenarios[itn, 0].util.first_stage_variables)
+        state_vars = list()
+        for blk in model.scenarios[itn, :]:
+            ssv_set = ComponentSet(blk.util.second_stage_variables)
+            state_vars.extend(
+                    v for v in blk.component_data_objects(Var,
+                                                          active=True,
+                                                          descend_into=True)
```
[the remark this answers is not in the document]
Comment (shermanjasonaf) [reply pyomo/contrib/pyros/master_problem_methods.py]: Noted; addressed in subsequent commits.

File: pyomo/contrib/pyros/master_problem_methods.py [hunk 2]
```diff
@@ -37,35 +41,186 @@ def initial_construct_master(model_data):
...
+                ssv_in_dr_eq.set_value(
+                        value(replace_expressions(eq.body,
+                                                  {id(ssv_in_dr_eq): 0}))
+                )
+
+    # initialize state vars to previous master solution values
+    if iteration != 0:
+        stvar_map = get_state_vars(model, [iteration, iteration-1])
+        # is order of variables in each state vars list guaranteed to
+        # be the same across blocks? if not, how to address that?
```
[the remark this answers is not in the document]
Comment (shermanjasonaf) [reply pyomo/contrib/pyros/master_problem_methods.py]: Noted; addressed in subsequent commits.

File: pyomo/contrib/pyros/master_problem_methods.py [hunk 3]
```diff
@@ -37,35 +41,186 @@ def initial_construct_master(model_data):
...
+
+                # get second-stage var in DR eqn. should only be one var
+                ssv_in_dr_eq = [var for var in vars_in_dr_eq
+                                if var in ssv_set][0]
+
+                # update var value for initialization
+                # fine since DR eqns are f(d) - z == 0 (not z - f(d) == 0)
+                ssv_in_dr_eq.set_value(
+                        value(replace_expressions(eq.body,
+                                                  {id(ssv_in_dr_eq): 0}))
```
[the remark this answers is not in the document]
Comment (shermanjasonaf) [reply pyomo/contrib/pyros/master_problem_methods.py]: Agreed. See subsequent commits.

**Pull request merged**